OrbitSshQt/include/OrbitSshQt/Base.h

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_SSH_QT_BASE_H_
+#define ORBIT_SSH_QT_BASE_H_
+
+#include <QObject>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitSsh/Error.h"
+#include "OrbitSshQt/Error.h"
+
+namespace OrbitSshQt {
+
+template <typename Derived, typename State_>
+class Base : public QObject {
+ public:
+  using State = State_;
+  using QObject::QObject;
+
+  void OnEvent() {
+    if (self()->currentState() > State::kInitial &&
+        self()->currentState() < State::kStarted) {
+      const auto result = self()->startup();
+
+      if (result) {
+        self()->started();
```

Comment:
The whole point of CTRE is to avoid virtual functions in the first place. If the user does not define started it will not compile. So I think it's fine.